File: python/mock_data/test_main.py
from pathlib import Path

import pytest
from faker import Faker
from main import (
    convert_unicode_to_ascii,
    generate_fake_persons,
    get_locations,
    write_persons_to_csv,
)

AGE_LOWER, AGE_UPPER = 22, 65


@pytest.fixture
def faker():
    return Faker()


@pytest.fixture
def location_data():
    locations = get_locations(Path("./data/worldcities.csv"))
    return locations


@pytest.mark.parametrize(
    "test_input,expected",
    [
        ("São Paulo", "Sao Paulo"),
        ("İstanbul", "Istanbul"),
        ("Mahārāshtra", "Maharashtra"),
        ("Středočeský Kraj", "Stredocesky Kraj"),
        ("Dar‘ā", "Dara"),
        ("Île-de-France", "Ile-de-France"),
    ],
)
def test_convert_unicode_to_ascii(test_input, expected):
    assert convert_unicode_to_ascii(test_input) == expected


def test_get_locations(location_data):
    assert len(location_data) == 44691


def test_generate_fake_persons(faker, location_data):
    profiles = generate_fake_persons(faker, location_data, 10)
    for profile in profiles:
        assert profile["id"] > 0
        assert len(profile["name"].split()) > 1
        assert AGE_LOWER <= profile["age"] <= AGE_UPPER


def test_write_persons_to_csv(faker, location_data):
    profiles = generate_fake_persons(faker, location_data, 10)
    output_path = Path("./data/test_persons.csv")
    write_persons_to_csv(profiles, output_path)
    assert output_path.exists()
    # Delete the file
    Path("./data/test_persons.csv").unlink()

File: rust/mock_data/src/test_main.rs
#![cfg(test)]

use std::path::Path;

#[test]
fn test_convert_unicode_to_ascii() {
    let name_list = [
        ("São Paulo", "Sao Paulo"),
        ("İstanbul", "Istanbul"),
        ("Mahārāshtra", "Maharashtra"),
        ("Středočeský Kraj", "Stredocesky Kraj"),
        ("Dar‘ā", "Dara"),
        ("Île-de-France", "Ile-de-France"),
    ];
    for (unicode, ascii) in name_list.iter() {
        assert_eq!(super::convert_unicode_to_ascii(unicode), *ascii);
    }
}

#[test]
fn test_read_cities() {
    let path = Path::new("data/worldcities.csv");
    let cities = super::read_cities(&path);
    assert_eq!(cities.len(), 44691);
}

#[test]
fn test_construct_person() {
    let age_lower = 18;
    let age_upper = 65;
    let path = Path::new("data/worldcities.csv");
    let locations = super::read_cities(&path);
    let person = super::construct_person(&locations, 1);
    assert!(person.id > 0);
    assert!(person.name.split(" ").count() > 1);
    assert!(age_lower <= person.age && person.age <= age_upper);
}